include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -1,405 +1,969 @@
-  static bool compute(
-    out IsotropicMicrofacetCache retval,
-    const surface_interactions::Isotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta
-  )
-  {
-    float3 dummy;
-    return nbl_glsl_calcIsotropicMicrofacetCache(_cache,transmitted,V,L,interaction.N,NdotL,VdotL,orientedEta,rcpOrientedEta,dummy);
-  }
-
-  bool isValidVNDFMicrofacet(const bool is_bsdf, const bool transmission, const float VdotL, const float eta, const float rcp_eta)
-  {
-    return NdotH >= 0.0 && !(is_bsdf && transmission && (VdotL > -min(eta,rcp_eta)));
-  }
-
-  float VdotH;
-  float LdotH;
-  float NdotH;
-  float NdotH2;
+        this_t retval;
+        retval.VdotL = VdotL;
+        retval.VdotH = LplusV_rcpLen * VdotL + LplusV_rcpLen;
+        retval.LdotH = retval.VdotH;
+        retval.NdotH = (NdotL + NdotV) * LplusV_rcpLen;
+        retval.NdotH2 = retval.NdotH * retval.NdotH;
+
+        return retval;
+    }
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL)
+    {
+        float dummy;
+        return createForReflection(NdotV, NdotL, VdotL, dummy);
+    }
+    template<class IsotropicInteraction, class LS NBL_FUNC_REQUIRES(surface_interactions::Isotropic<IsotropicInteraction> && LightSample<LS>)
+    static this_t createForReflection(
+        NBL_CONST_REF_ARG(IsotropicInteraction) interaction,
+        NBL_CONST_REF_ARG(LS) _sample)
+    {
+        return createForReflection(interaction.getNdotV(), _sample.getNdotL(), hlsl::dot<vector3_type>(interaction.getV().getDirection(), _sample.getL().getDirection()));
+    }
+
+    // transmissive cases need to be checked if the path is valid before usage
+    static this_t create(const bool transmitted, NBL_CONST_REF_ARG(ComputeMicrofacetNormal<scalar_type>) computeMicrofacetNormal, const scalar_type VdotL,
+        const vector3_type N, NBL_REF_ARG(vector3_type) H)
+    {
+        this_t retval;
+        retval.VdotL = VdotL;
+        H = computeMicrofacetNormal.normalized(transmitted);
+        retval.NdotH = hlsl::dot<vector3_type>(N, H);
+
+        // not coming from the medium (reflected) OR
+        // exiting at the macro scale AND ( (not L outside the cone of possible directions given IoR with constraint VdotH*LdotH<0.0) OR (microfacet not facing toward the macrosurface, i.e. non heightfield profile of microsurface) )
+        const bool valid = ComputeMicrofacetNormal<scalar_type>::isValidMicrofacet(transmitted, VdotL, retval.NdotH, computeMicrofacetNormal.orientedEta);
+        if (valid)
+        {
+            retval.VdotH = hlsl::dot<vector3_type>(computeMicrofacetNormal.V,H);
+            retval.LdotH = hlsl::dot<vector3_type>(computeMicrofacetNormal.L,H);
+            retval.NdotH2 = retval.NdotH * retval.NdotH;
+        }
+        else
+            retval.NdotH = bit_cast<scalar_type>(numeric_limits<scalar_type>::quiet_NaN);
```

Comment:
id `NdotH` is negative, we need to flip all dot products with `H` and H` itself, for why see https://github.com/Devsh-Graphics-Programming/Nabla/pull/899#discussion_r2244485044